Animate tree lookup and render

shreyass-ranganatha opened the pull request:
## Fixes
<!-- If PR doesn't fully resolve the issue, replace 'Fixes' below with 'Related to'. -->
<!-- If there is no issue being resolved, please consider opening one before creating this pull request. -->
Fixes #41 by @dhruvkb

## Description
- Made use of `rich.live.Live` to pretty print tree lookup and rendering procedures.
- Replaced the many recursive functions (`match`, `find_children`, `set_sub_pre_shapes`) that were 
used for the tree before to a single recursive block, that performs all the functions, without the 3 
separate loops
<!-- Concisely describe what the pull request does. -->
<!-- Add screenshots, videos, or other media to show the problem and the solution when appropriate. -->

## Testing Instructions
<!-- Give steps for the reviewer to verify that this PR fixes the problem; or delete this section entirely. -->
0. Check out the PR and run the app in development mode.

## Checklist
<!-- Replace  the [ ] with [x] to check the boxes. -->
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`).
- [x] My pull request targets the *default* branch of the repository (`main`) or a parent feature branch.
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [ ] I added or updated tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no visible errors.

[best_practices]:https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin
<!-- You must read and understand the following attestation. -->

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

dhruvkb commented on src/pls/models/node.py:
```diff
@@ -17,6 +17,7 @@
 from pls.models.composables.stat import StatComp
 from pls.models.composables.type import TypeComp
 from pls.models.format_rules import FormatRules
+from pls.models.node_spec import NodeSpec
```
This can be moved into the `if TYPE_CHECKING` block below as it's only used for type annotation.

dhruvkb commented on src/pls/models/composables/spec.py:
```diff
@@ -31,10 +31,12 @@ def match(self, specs: list[NodeSpec]):
-            for child in self.node.children:
-                if isinstance(child, type(self.node)):
-                    child.spec_comp.match(specs)
+
+        # Deprecated:
+        # if args.args.tree:
+        #     for child in self.node.children:
+        #         if isinstance(child, type(self.node)):
+        #             child.spec_comp.match(specs)
```
Feel free to delete this entirely rather than making it a comment.

dhruvkb reviewed (changes requested):
Thanks @shreyass-ranganatha! This is an awesome 👌 improvement.

The output looks pretty good and feels much more responsive compared to the previous behaviour which would show nothing for a while and then everything at once.

It'll be nice to prune the inaccessible codepaths rather than the "Deprecated" comments. If we need to revert a change we can do so from the Git history.

shreyass-ranganatha commented:
I've implemented the requests, and fixes an issue where the same directory and files used to render multiple times. 
Also fixed the bit where hidden files were shown without the `--all` parameter.
